bloptools/bayesian/dofs.py

```diff
@@ -0,0 +1,253 @@
+import time as ttime
+import uuid
+from collections.abc import Sequence
+from dataclasses import dataclass, field
+from typing import Tuple, Union
+
+import numpy as np
+import pandas as pd
+from ophyd import Signal, SignalRO
+
+DEFAULT_BOUNDS = (-5.0, +5.0)
+DOF_FIELDS = ["description", "readback", "lower_limit", "upper_limit", "units", "active", "read_only", "tags"]
+
+numeric = Union[float, int]
+
+
+class ReadOnlyError(Exception):
+    ...
+
+
+def _validate_dofs(dofs):
+    dof_names = [dof.name for dof in dofs]
+
+    # check that dof names are unique
+    unique_dof_names, counts = np.unique(dof_names, return_counts=True)
+    duplicate_dof_names = unique_dof_names[counts > 1]
+    if len(duplicate_dof_names) > 0:
+        raise ValueError(f'Duplicate name(s) in supplied dofs: "{duplicate_dof_names}"')
+
+    return list(dofs)
+
+
+@dataclass
+class DOF:
+    device: Signal = None
+    description: str = None
+    name: str = None
+    limits: Tuple[float, float] = (-10.0, 10.0)
+    units: str = ""
+    read_only: bool = False
+    active: bool = True
+    tags: list = field(default_factory=list)
+    latent_group: str = None
+
+    def __post_init__(self):
+        self.uuid = str(uuid.uuid4())
+
+        if self.name is None:
+            self.name = self.device.name if hasattr(self.device, "name") else self.uuid
+
+        if self.device is None:
+            self.device = Signal(name=self.name)
+
+        if not self.read_only:
+            # check that the device has a put method
+            if isinstance(self.device, SignalRO):
+                raise ValueError("Must specify read_only=True for a read-only device!")
+
+        if self.latent_group is None:
+            self.latent_group = str(uuid.uuid4())
+
+        # all dof degrees of freedom are hinted
+        self.device.kind = "hinted"
+
+    @property
+    def lower_limit(self):
+        return float(self.limits[0])
+
+    @property
+    def upper_limit(self):
+        return float(self.limits[1])
+
+    @property
+    def readback(self):
+        return self.device.read()[self.device.name]["value"]
+
+    @property
+    def summary(self) -> pd.Series:
+        series = pd.Series(index=DOF_FIELDS)
+        for attr in series.index:
+            series[attr] = getattr(self, attr)
+        return series
+
+    @property
+    def label(self) -> str:
+        return f"{self.name}{f' [{self.units}]' if len(self.units) > 0 else ''}"
+
+    @property
+    def has_model(self):
+        return hasattr(self, "model")
+
+
+class DOFList(Sequence):
+    def __init__(self, dofs: list = []):
+        _validate_dofs(dofs)
+        self.dofs = dofs
+
+    def __getitem__(self, i):
+        if type(i) is int:
+            return self.dofs[i]
+        elif type(i) is str:
+            return self.dofs[self.names.index(i)]
```

Comment:
The `else` case with an exception raised would help here to avoid silent `None` returns.